File: 2023/UICPC DIV1/UICPC Round #01/solutions/C/david.java
import java.util.ArrayDeque;
import java.util.Scanner;

public class david {
	private final static int[] dx = {1,0,-1,0};
	private final static int[] dy = {0,1,0,-1};
	
	private static boolean isPossible(int R, int C, int K, int H, int[][] grid) {
		boolean done[][] = new boolean[R][C];
		
		ArrayDeque<Integer> work = new ArrayDeque<Integer>();
		ArrayDeque<Integer> nextWork = new ArrayDeque<Integer>();
		System.err.println("S");
		for (int i=0; i<=K; i++) {
			if (i == 0) {
				// setup
				for (int j=0; j<C; j++) {
					if (grid[i][j] >= H)
						work.add(j);
					else
						nextWork.add(j);
					done[0][j] = true;
				}
			}
			
			while (!work.isEmpty()) {
				int pos = work.remove();
				int x = pos/C;
				int y = pos%C;
				
				if (x == R-1) return true;
				for (int j=0; j<4; j++) {
					if (x + dx[j] < 0 || x + dx[j] >= R || y + dy[j] < 0 || y + dy[j] >= C) continue;
					if (done[x+dx[j]][y+dy[j]]) continue;
					if (grid[x+dx[j]][y+dy[j]] >= H)
						work.add((x+dx[j])*C+y+dy[j]);
					else
						nextWork.add((x+dx[j])*C+y+dy[j]);
					done[x+dx[j]][y+dy[j]] = true;
				}
			}
			
			ArrayDeque<Integer> t = work;
			work = nextWork;
			nextWork = t;
		}
		
		return false;
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		
		int R = sc.nextInt();
		int C = sc.nextInt();
		int K = sc.nextInt();
		
		int grid[][] = new int[R][C];
		for (int i=0; i<R; i++) {
			for (int j=0; j<C; j++) {
				grid[i][j] = sc.nextInt();
			}
		}
		
		int low = 0;
		int high = 1000000001;
		while (high - low > 1) {
			int mid = (high+low)/2;
			if (isPossible(R,C,K, mid, grid))
				low = mid;
			else
				high = mid;
		}
		
		System.out.println(low);
	}
}

File: 2023/UICPC DIV1/UICPC Round #01/solutions/C/david.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

bool isPossible(int H, vector<vector<int>> const &canyon, int K) {
	vector<vector<bool>> done(canyon.size(), vector<bool>(canyon[0].size(), false));
	
	queue<pair<int, int>> work;
	for (int i=0; i<=K; i++) {
		queue<pair<int, int>> nextWork;
		if (i == 0) {
			// Setup
			for (int j=0; j<(int)canyon[0].size(); j++) {
				// Add to correct queue
				if (canyon[0][j] >= H)
					work.push({0,j});
				else
					nextWork.push({0,j});
				done[0][j] = true; // and mark
			}
		}
		
		while (!work.empty()) {
			auto pos = work.front();
			work.pop();
			
			if (pos.first == (int)canyon.size()-1) return true; // we are through
			
			vector<pair<int, int>> d = {{0,1},{1,0},{-1,0},{0,-1}};
			for (auto delta : d) {
				int x = pos.first + delta.first;
				int y = pos.second + delta.second;
				
				// Valid pos?
				if (x < 0 || y < 0 || x >= (int)canyon.size() || y >= (int)canyon[0].size()) continue;
				
				// Already visited?
				if (done[x][y]) continue;
				
				// Add to correct queue
				if (canyon[x][y] >= H)
					work.push({x,y});
				else
					nextWork.push({x,y});
				done[x][y] = true; // and mark
			}
		}
		
		// Use a bridge and continue
		work.swap(nextWork);
	}
	
	// no possible path
	return false;
}

int main() {
	// Read in testdata
	int R, C, K;
	cin >> R >> C >> K;
	vector<vector<int>> canyon(R, vector<int>(C));
	for (int i=0; i<R; i++) {
		for (int j=0; j<C; j++) {
			cin >> canyon[i][j];
		}
	}
	
	// Binary search for the largest possible path height
	int low = 0;
	int high = 1000000001;
	while (high - low > 1) {
		int mid = (high+low)/2;
		if (isPossible(mid, canyon, K))
			low = mid;
		else
			high = mid;
	}
	
	// And output results
	cout << low << endl;
}
